codebasin/report.py

```diff
@@ -81,6 +82,73 @@ def divergence(setmap):
     return d / float(npairs)
 
 
+def utilization(setmap: defaultdict[frozenset[str], int]) -> float:
+    """
+    Compute the average code utilization for all lines in the setmap.
+    i.e., (reused SLOC / total SLOC)
+
+    Parameters
+    ----------
+    setmap: defaultdict[frozenset[str], int]
+        The mapping from platform sets to SLOC.
+
+    Returns
+    -------
+    float
+        The average code utilization, in the range [0, NumPlatforms].
+    """
+    reused_sloc = 0
+    total_sloc = 0
+    for k, v in setmap.items():
+        reused_sloc += len(k) * v
+        total_sloc += v
+    if total_sloc == 0:
+        return float("nan")
+
+    return reused_sloc / total_sloc
+
+
+def normalized_utilization(
+    setmap: defaultdict[frozenset[str], int],
+    total_platforms: int | None = None,
+) -> float:
+    """
+    Compute the average code utilization, normalized for a specific number of
+    platforms.
+
+    Parameters
+    ----------
+    setmap: defaultdict[frozenset[str,int]
+        The mapping from platform sets to SLOC.
+
+    total_platforms: int | None, default: None
+        The total number of platforms to use as the denominator.
+        If None, the denominator will be derived from the setmap.
+
+    Returns
+    -------
+    float
+        The average code utilization, in the range [0, 1].
+
+    Raises
+    ------
+    ValueError
+        If `total_platforms` < the number of platforms in `setmap`.
```

Comment:
If you're formatting for `sphinx` autodoc in mind, I think it should be double marks (``total_platforms``) right? Disregard if you don't care

Reply:
I actually didn't know about this, so thanks for pointing it out.

I think we don't need to worry about it, though, because we're not writing in the `sphinx` autodoc format.  We're using NumPy style docstrings, which get run through a different `sphinx` extension (i.e., [napoleon](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html)).  NumPy docstrings expect a single ` for parameters, and this seems to be working okay in the P3 Analysis Library configuration.  My guess is that napoleon is adding the extra backticks for us, or something like that.